### PR TITLE
Switch to scitokens-cpp implementation

### PR DESCRIPTION
**Pull request by bbockelm**
This PR switches the xrootd plugin from the Python implementation of scitokens to the `scitokens-cpp` one.

### CODE REVIEW

**Comment by bbockelm** [comment]
Looks like we need `scitokens-cpp` in OSG dev repo in order for the Travis-CI tests to have a hope of passing...

File: src/scitokens.cpp [hunk 1]
```diff
@@ -207,34 +276,347 @@ class XrdAccSciTokens : public XrdAccAuthorize
...
+    bool GenerateAcls(const std::string &authz, uint64_t &cache_expiry, AccessRulesRaw &rules, std::string &username) {
+        if (strncmp(authz.c_str(), "Bearer%20", 9)) {
+            return false;
+        }
+
+        // Does this look like a JWT?  If not, bail out early and
+        // do not pollute the log.
+        bool looks_good = true;
+        int separator_count = 0;
+        for (auto cur_char = authz.c_str() + 7; *cur_char; cur_char++) {
```
Comment (djw8605) [inline src/scitokens.cpp]: Why 7 and not 9?  Isn't 9 the start of the bearer token (you used 9 above when testing for `Bearer%20`)

File: src/scitokens.cpp [hunk 2]
```diff
@@ -207,34 +276,347 @@ class XrdAccSciTokens : public XrdAccAuthorize
...
+        if (config.m_map_subject) {
+            value = nullptr;
+            if (scitoken_get_claim_string(token, "sub", &value, &err_msg)) {
+                pthread_rwlock_unlock(&m_config_lock);
+                m_log.Emsg("GenerateAcls", "Failed to get token subject:", err_msg);
+                free(err_msg);
+                scitoken_destroy(token);
+                return false;
+            }
+            token_username = std::string(value);
```
Comment (djw8605) [inline src/scitokens.cpp:397]:
```suggestion
            token_username = std::string(value);
            free(value);
```

**Review by djw8605 — CHANGES_REQUESTED** [review]
A few comments.  Remove python, and add over 100 lines of code.  Probably still 1000x faster.

All of the pathing logic that is required sucks.  I wish there was a way to handle that more gracefully.

File: src/scitokens.cpp (hunk 1 above)
Comment (bbockelm) [reply src/scitokens.cpp]: You're correct, the offset is 9.

**Review by djw8605 — APPROVED** [review]
LGTM

**Comment by matyasselmeci** [comment]
I put scitokens-cpp into osg-development; can you rerun the travis tests to see if they pass?

**Comment by bbockelm** [comment]
@djw8605 - I'm in an all-day meeting today.  Can you push the corresponding fixes?

I think we need to tweak the CI script so it does an explicit `yum install` of `scitokens-cpp`.

**Comment by bbockelm** [comment]
Whoops - scitokens-cpp-devel

**Comment by djw8605** [comment]
A few minutes after I committed, I realized my mistake.  Rebuilding.

**Comment by matyasselmeci** [comment]
Looks like you need to put the submodules into the source tarball you make.  In scitokens-cpp you used git_archive_all.py from https://github.com/Kentzo/git-archive-all to do that.

**Comment by djw8605** [comment]
Requires the underlying scitokens-cpp fix: https://github.com/scitokens/scitokens-cpp/pull/8

**Comment by djw8605** [comment]
Tests are green.  Merging

**Pull request merged**